syntax = "proto3";

package magda.mcp;

option cpp_namespace = "magda::mcp";

// Multi-agent Control Protocol Service
service MCPService {
    // Execute a single command
    rpc ExecuteCommand(CommandRequest) returns (CommandResponse);
    
    // Stream of bidirectional communication
    rpc CommandStream(stream CommandRequest) returns (stream CommandResponse);
    
    // Register an agent
    rpc RegisterAgent(RegisterAgentRequest) returns (RegisterAgentResponse);
    
    // Send a message to specific agent
    rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
    
    // Broadcast message to all agents
    rpc BroadcastMessage(BroadcastMessageRequest) returns (BroadcastMessageResponse);
    
    // Get connected agents
    rpc GetConnectedAgents(GetConnectedAgentsRequest) returns (GetConnectedAgentsResponse);
}

// Parameter value (equivalent to our Command::ParamValue)
message ParameterValue {
    oneof value {
        string string_value = 1;
        int32 int_value = 2;
        double double_value = 3;
        bool bool_value = 4;
        DoubleArray double_array_value = 5;
    }
}

message DoubleArray {
    repeated double values = 1;
}

// Command message
message CommandRequest {
    string command_type = 1;
    map<string, ParameterValue> parameters = 2;
    string agent_id = 3;  // Optional: for identifying the sender
}

// Response message
message CommandResponse {
    enum Status {
        SUCCESS = 0;
        ERROR = 1;
        PENDING = 2;
    }
    
    Status status = 1;
    string message = 2;
    string data = 3;  // JSON string for additional data
}

// Agent registration
message RegisterAgentRequest {
    string agent_name = 1;
    string agent_type = 2;  // e.g., "composer", "mixer", "analyzer"
    map<string, string> capabilities = 3;
}

message RegisterAgentResponse {
    string agent_id = 1;
    string server_version = 2;
    CommandResponse status = 3;
}

// Message sending
message SendMessageRequest {
    string target_agent_id = 1;
    string message = 2;  // JSON message
}

message SendMessageResponse {
    CommandResponse status = 1;
}

// Broadcasting
message BroadcastMessageRequest {
    string message = 1;  // JSON message
    string sender_agent_id = 2;  // Optional: exclude sender from broadcast
}

message BroadcastMessageResponse {
    CommandResponse status = 1;
    int32 recipients_count = 2;
}

// Get connected agents
message GetConnectedAgentsRequest {
    // Empty - just requesting the list
}

message GetConnectedAgentsResponse {
    repeated AgentInfo agents = 1;
}

message AgentInfo {
    string agent_id = 1;
    string agent_name = 2;
    string agent_type = 3;
    int64 connected_timestamp = 4;  // Unix timestamp
    map<string, string> capabilities = 5;
} 